R/toolISOhistorical.R

```diff
@@ -115,56 +116,253 @@ toolISOhistorical <- function(m, mapping = NULL, additional_mapping = NULL, over
+          } else if (mainDim == 1.2) {
+            addInd <- setdiff(as.vector(outer(
+              getItems(z[setNames(list(c(tr$fromISO)), mainDimName), subTime, ], dim = secdDim),
+              tr$toISO, paste, sep = ".")),
+              getItems(weight, dim = 1))
+          }
+          if (length(addInd) > 0) {
+            addR <- unique(c(addR, addInd))
+          }
+          if (length(addR) > 0) {
+            addMiss <- new.magpie(cells_and_regions =  addR,
+                                  years = getItems(weight, dim = 2),
+                                  names = getItems(weight, dim = 3),
+                                  fill = 0)
+            weight <- mbind(weight, addMiss)
+          }
+
+
+          if (anyNA(weight)) {
+            weight[is.na(weight)] <- 0
+            vcat(0, "Weight 1 in toolISOhistorical contained NAs. Set NAs to 0!")
+          }
+        } else {
+          stop("additional_weight is not supported for bilateral data.")
         }
-        weight <- additional_weight[a$toISO, , ]
+
+        bilatMapping <- mapping # mapping needs to be made bilateral
+        if (mainDim == 1.1) {
+bilatMapping <- data.frame(toISO = as.vector(outer(bilatMapping$toISO, getItems(z, dim = secdDim), paste, sep = ".")),
+                                                        fromISO = as.vector(outer(bilatMapping$fromISO,
+                                                                  getItems(z, dim = secdDim), paste, sep = ".")))
+        } else if (mainDim == 1.2) {
+bilatMapping <- data.frame(toISO = as.vector(outer(getItems(z, dim = secdDim), bilatMapping$toISO, paste, sep = ".")),
+                                                              fromISO = as.vector(outer(getItems(z, dim = secdDim),
+                                                                        bilatMapping$fromISO, paste, sep = ".")))
+        }
+
+        # make a second weight to add to original weight in cases where weight is 0 in all new countries.
+        # This weight is very small value so should not affect actual values. Weight2 is ratio of each country's values
+        # as ratio of sum across  partner countries as well as products.
+        weight2 <- weight
+        for (i in getItems(weight2, dim = mainDim)) {
+          weight2[i, , ] <- dimSums(weight2[i, , ], dim = c(secdDim, 3)) /
+                            dimSums(weight2[i, , ], dim = c(1, 3), na.rm = TRUE)
+        }
+        weight2[is.na(weight2)] <- 0
+        weight2 <- weight2 / 1e10
+
+        zTr <- toolAggregate(z[setNames(list(c(tr$fromISO)), mainDimName), subTime, ],
+                             bilatMapping[is.element(bilatMapping$toISO,
+                                                     getItems(weight[setNames(list(c(getItems(z[setNames(list(c(tr$fromISO)), #nolint
+                                                                                                         mainDimName), , ], dim = secdDimName))),  # nolint
```

Comment:
These `# nolint` are because of line length right? I'd strongly suggest to split this up into multiple statements, I find it very hard to read and understand it like this.